internal/persistence/sql/relationtuples.go

```diff
@@ -18,105 +20,210 @@ import (
 	}
 
-	n, err := p.namespaces.GetNamespace(ctx, rel.Namespace)
+	p.d.Logger().WithFields(rel.ToLoggerFields()).Trace("creating in database")
+
+	rt := &RelationTuple{
+		ID:         uuid.Must(uuid.NewV4()),
+		NetworkID:  p.NetworkID(ctx),
+		CommitTime: time.Now(),
+	}
+	if err := rt.fromInternal(ctx, p, rel); err != nil {
+		return err
+	}
+
+	if err := sqlcon.HandleError(
+		p.Connection(ctx).Create(rt),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Persister) deleteRelationTupleSubjectID(ctx context.Context, r *relationtuple.InternalRelationTuple, s *relationtuple.SubjectID, n *namespace.Namespace, network uuid.UUID) error {
+	if err := p.Connection(ctx).RawQuery(
+		"DELETE FROM keto_relation_tuples WHERE nid = ? AND namespace_id = ? AND object = ? AND relation = ? AND subject_id = ?",
+		network,
+		n.ID,
+		r.Object,
+		r.Relation,
+		s.ID,
+	).Exec(); err != nil {
+		return sqlcon.HandleError(err)
+	}
+
+	return nil
+}
+
+func (p *Persister) deleteRelationTupleSubjectSet(ctx context.Context, r *relationtuple.InternalRelationTuple, s *relationtuple.SubjectSet, n *namespace.Namespace, network uuid.UUID) error {
+	nm, err := p.d.Config().NamespaceManager()
 	if err != nil {
 		return err
 	}
 
-	p.l.WithFields(rel.ToLoggerFields()).Trace("creating in database")
+	sn, err := nm.GetNamespaceByName(ctx, s.Namespace)
+	if err != nil {
+		return err
+	}
 
-	return sqlcon.HandleError(
-		p.connection(context.WithValue(ctx, namespaceContextKey, n)).Create(&relationTuple{
-			ID:         shardID,
-			Object:     rel.Object,
-			Relation:   rel.Relation,
-			Subject:    rel.Subject.String(),
-			CommitTime: time.Now(),
-		}),
-	)
+	if err := p.Connection(ctx).RawQuery(
+		"DELETE FROM keto_relation_tuples WHERE nid = ? AND namespace_id = ? AND object = ? AND relation = ? AND subject_set_namespace_id = ? AND subject_set_object = ? AND subject_set_relation = ?",
+		network,
+		n.ID,
+		r.Object,
+		r.Relation,
+		sn.ID,
+		s.Object,
+		s.Relation,
+	).Exec(); err != nil {
+		return sqlcon.HandleError(err)
+	}
+	return nil
 }
 
 func (p *Persister) DeleteRelationTuples(ctx context.Context, rs ...*relationtuple.InternalRelationTuple) error {
+	nm, err := p.d.Config().NamespaceManager()
+	if err != nil {
+		return err
+	}
+
 	return p.transaction(ctx, func(ctx context.Context, c *pop.Connection) error {
 		for _, r := range rs {
-			if r.Subject == nil {
-				return errors.New("subject is not allowed to be nil")
-			}
-
-			n, err := p.namespaces.GetNamespace(ctx, r.Namespace)
+			n, err := nm.GetNamespaceByName(ctx, r.Namespace)
 			if err != nil {
 				return err
 			}
 
-			if err := c.RawQuery(
-				fmt.Sprintf("DELETE FROM %s WHERE object = ? AND relation = ? AND subject = ?", tableFromNamespace(n)),
-				r.Object,
-				r.Relation,
-				r.Subject.String(),
-			).Exec(); err != nil {
-				return sqlcon.HandleError(err)
+			switch s := r.Subject.(type) {
+			case *relationtuple.SubjectID:
```

Comment:
Would it maybe make sense to find the element, and then delete it? So that we have only one place which constructs the (quite) complex search query?

Reply:
What do you mean by find it? Use `GetRelationTuples` and then use the `shard_id` for deletion instead? How do you then handle duplicate tuples? Get all of the matching tuples, maybe even paginated, and delete with `shard_id = ? OR shard_id = ? ...`?

Reply:
Ah I see - I forgot about duplication :)